lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3670,40 +3670,114 @@ void TypeChecker::checkParameterList(ParameterList *params,
+    const auto result = TypeResolution::resolveContextualType(
+        specializationArg, dc, options,
+        // Introduce type variables for unbound generics.
+        OpenUnboundGenericType(cs, locator),
+        HandlePlaceholderType(cs, locator),
+        OpenPackElementType(cs, locator, nullptr));
+    if (result->hasError())
+      return true;
+
+    specializationArgTypes.push_back(result);
+  }
+  cs.addConstraint(
+      ConstraintKind::ExplicitGenericArguments, boundType,
+      PackType::get(cs.getASTContext(), specializationArgTypes),
+      locator);
+  return false;
+}
+
+BraceStmt *
 ExpandMacroExpansionDeclRequest::evaluate(Evaluator &evaluator,
                                           MacroExpansionDecl *MED) const {
   auto &ctx = MED->getASTContext();
   auto *dc = MED->getDeclContext();
-  auto foundMacros = TypeChecker::lookupMacros(
-      MED->getDeclContext(), MED->getMacro(),
-      MED->getLoc(), MacroRole::Declaration);
+  // Non-call declaration macros cannot be overloaded.
+  MacroRoles viableRoles(MacroRole::Declaration);
+  // In a code block context, a macro expansion decl can also expand to an
+  // arbitrary code item.
+  if (dc->isLocalContext() ||
+      dc->getContextKind() == DeclContextKind::FileUnit) {
+    viableRoles |= MacroRole::Expression;
+    viableRoles |= MacroRole::CodeItem;
+  }
+  SmallVector<MacroDecl *, 4> foundMacros;
+  auto foundRoles = TypeChecker::lookupMacros(
+      MED->getDeclContext(), MED->getMacro(), MED->getLoc(), viableRoles,
+      foundMacros);
   if (foundMacros.empty()) {
     MED->diagnose(diag::macro_undefined, MED->getMacro().getBaseIdentifier())
         .highlight(MED->getMacroLoc().getSourceRange());
-    return {};
+    return nullptr;
   }
+
+  auto expandAsMacroExpansionExpr = [&]() -> BraceStmt * {
+    Expr *mee = MED->createExpr();
+    TypeChecker::typeCheckExpression(mee, dc, {},
+                                     TypeCheckExprFlags::IsExprStmt |
+                                         TypeCheckExprFlags::IsDiscarded);
+    ASTNode expansion = mee;
+    // If top-level, wrap it in a `TopLevelCodeDecl(BraceStmt(...))`.
+    auto innermostDC = dc;
+    if (dc->getContextKind() == DeclContextKind::FileUnit) {
+      auto *tlcd = new (ctx) TopLevelCodeDecl(
+          dc, BraceStmt::createImplicit(ctx, {expansion}));
+      tlcd->setDeclContext(dc);
+      expansion = tlcd;
+      innermostDC = tlcd;
+    }
+    auto codeBlockItemList = BraceStmt::createImplicit(ctx, {expansion});
+    MED->setRewritten(codeBlockItemList);
+    TypeChecker::postTypeCheckMacroExpansion(mee, innermostDC);
+    return codeBlockItemList;
+  };
+
+  // If we've only found expression macros, expand to a macro expansion expr.
+  if (foundRoles.containsOnly(MacroRole::Expression))
+    return expandAsMacroExpansionExpr();
+
   // Resolve macro candidates.
   MacroDecl *macro;
-  // Non-call declaration macros cannot be overloaded.
+  // Non-call macros cannot be overloaded.
```

Comment:
I don't think that's necessarily true, but it can be follow-up work to let the constraint system sort it out.